# Практическая работа №2: Обработка выборочных данных. Нахождение точечных оценок параметров распределения

Выполнили студенты гр. 2384 Исмаилов Максим Владимирович и Дамакин Роман Павлович. Вариант №31.

## Цель работы

Освоение практических навыков по нахождению точечных статистических оценок параметров распределения.

## Основные теоретические положения

**Метод условных вариант (метод произведений)** применяется для упрощения вычислений числовых характеристик по сгруппированным данным. Вводятся условные варианты:
$$ u_i = \frac{x_i - C}{h} $$
где $C$ — «ложный нуль» (часто выбирают середину интервала с максимальной частотой), а $h$ — ширина (длина) интервала.

Условные эмпирические моменты $k$-го порядка определяются так:
$$ \bar{M}^*_k = \frac{1}{n} \sum_{i=1}^{k} m_i u_i^k $$

Через условные моменты выражаются основные выборочные характеристики:
$$ \bar{x}_{\textit{в}} = \bar{M}^*_1 \, h + C $$
$$ D_{\textit{в}} = \left(\bar{M}^*_2 - (\bar{M}^*_1)^2\right) h^2 $$

---

**Выборочное среднее** — это среднее арифметическое значений выборки (с учётом частот):
$$ \bar{x}_{\textit{в}} = \frac{1}{n} \sum_{i=1}^{k} m_i x_i $$

**Выборочная дисперсия** характеризует разброс наблюдений относительно выборочного среднего:
$$ D_{\textit{в}} = \frac{1}{n} \sum_{i=1}^{k} m_i (x_i - \bar{x}_{\textit{в}})^2 = \overline{x^2} - (\bar{x}_{\textit{в}})^2 $$

**Выборочное среднеквадратическое отклонение (СКО)** вычисляется как корень из дисперсии:
$$ \sigma_{\textit{в}} = \sqrt{D_{\textit{в}}} $$

---

Поскольку выборочная дисперсия $D_{\textit{в}}$ является смещённой оценкой генеральной дисперсии, используют **исправленную (несмещённую) дисперсию**:
$$ s^2 = \frac{n}{n-1} D_{\textit{в}} $$
Соответствующее **исправленное СКО**:
$$ s = \sqrt{s^2} $$

---

Центральные эмпирические моменты $ \bar{m}_k $ можно выразить через условные моменты:
$$ \bar{m}_2 = \left(\bar{M}^*_2 - (\bar{M}^*_1)^2\right) h^2 $$
$$ \bar{m}_3 = \left(\bar{M}^*_3 - 3\bar{M}^*_2\bar{M}^*_1 + 2(\bar{M}^*_1)^3\right) h^3 $$
$$ \bar{m}_4 = \left(\bar{M}^*_4 - 4\bar{M}^*_3\bar{M}^*_1 + 6\bar{M}^*_2(\bar{M}^*_1)^2 - 3(\bar{M}^*_1)^4\right) h^4 $$

---

**Коэффициент асимметрии** $ \bar{A}_s $ отражает степень и направление скошенности распределения:
$$ \bar{A}_s = \frac{\bar{m}_3}{\sigma_{\textit{в}}^3} $$
* $ \bar{A}_s > 0 $ — асимметрия вправо.
* $ \bar{A}_s < 0 $ — асимметрия влево.

**Эксцесс** $ \bar{E} $ описывает степень островершинности распределения (по сравнению с нормальным):
$$ \bar{E} = \frac{\bar{m}_4}{\sigma_{\textit{в}}^4} - 3 $$
* $ \bar{E} > 0 $ — распределение более островершинное.
* $ \bar{E} < 0 $ — распределение более плосковершинное.

---

**Мода** $M^*_o$ — наиболее часто встречающееся значение (для интервального ряда — оценка по модальному интервалу):
$$ M^*_o = x_{Mo}^{(0)} + h \frac{m_{Mo} - m_{Mo-1}}{(m_{Mo} - m_{Mo-1}) + (m_{Mo} - m_{Mo+1})} $$
где $x_{Mo}^{(0)}$ — нижняя граница модального интервала.

**Медиана** $M^*_e$ — значение, делящее вариационный ряд на две равные по числу наблюдений части:
$$ M^*_e = x_{Me}^{(0)} + h \frac{0.5n - S_{Me-1}}{m_{Me}} $$
где $x_{Me}^{(0)}$ — нижняя граница медианного интервала, а $S_{Me-1}$ — накопленная частота интервала перед медианным.

---

**Коэффициент вариации** показывает относительный уровень разброса:
$$ V^* = \frac{\sigma_{\textit{в}}}{\bar{x}_{\textit{в}}} \cdot 100\% $$
* Если $V^* \le 33\%$, совокупность обычно считают однородной.

## Постановка задачи

По заданной выборке, используя метод моментов и метод условных вариантов, найти точечные статистические оценки: математического ожидания, дисперсии, среднеквадратического отклонения, коэффициента асимметрии, эксцесса, моды, медианы и коэффициента вариации исследуемой случайной величины. Затем дать содержательную интерпретацию полученных результатов.


## Порядок выполнения работы

1. Для середин интервального ряда, полученного в практической работе №1, вычислить условные варианты. Заполнить табл. 1 (в последней строке
Σ необходимо заполнить суммы столбцов; ячейки отмеченные прочерком заполнять не надо). Провести контроль вычислений.
2. Вычислить условные эмпирические моменты $\bar{M}_k^{*}$ через условные варианты. С помощью условных эмпирических моментов вычислить центральные эмпирические моменты $\bar{m}_k$. Полученные результаты занести в табл. 2.
3. Вычислить выборочные среднее $\bar{x}_в$ и дисперсию $D_в$ с помощью стандартной формулы и с помощью условных вариантов. Убедиться, что результаты совпадают. Вычислить выборочное СКО $σ_в$.
4. Вычислить исправленную выборочную дисперсию $s^2$ и исправленное $СКО s$. Сравнить данные оценки с смещёнными оценками дисперсии и СКО. Сделать выводы.
5. Найти статистическую оценку коэффициентов асимметрии $\bar{A}_s$ и эксцесса
$\bar{E}$. Сделать выводы.
6. Для интервального ряда вычислить моду $M_o^{*}$, медиану $M_e^{*}$ и коэффициент вариации $V^{*}$ заданного распределения. Сделать выводы.

## Выполнение работы

In [213]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from IPython.display import display, Markdown

def md(df: pd.DataFrame) -> None:
    display(Markdown(df.to_markdown(index=False)))

def _no_sigma(df: pd.DataFrame) -> pd.DataFrame:
    # Σ может быть строкой/числом — приводим к строке и режем
    return df[df["i"].astype(str) != "Σ"].copy()

def _pick(m: dict, *keys, name: str = ""):
    for k in keys:
        if k in m:
            return m[k]

def _parse_interval(s: str) -> tuple[float, float]:
    # "[a, b)" или "[a, b]" -> (a,b)
    s = str(s).strip()
    s = s.lstrip("[").rstrip(")]").rstrip("]")
    left, right = s.split(",")
    return float(left), float(right)

# Загрузка данных:

csv_path = "sample.csv"

df_raw = pd.read_csv(csv_path, skiprows=3)

df_raw.head()

,nu,E
0,480,153.3
1,510,129.4
2,426,119.0
3,482,139.9
4,393,103.2


In [214]:
def interval_table_hist(x: np.ndarray, n: int):
    k = int(round(1 + 3.322 * np.log10(n)))

    x_min, x_max = float(x.min()), float(x.max())
    edges = np.linspace(x_min, x_max, k + 1)
    counts, bin_edges = np.histogram(x, bins=edges)  # считает частоты

    mids = (bin_edges[:-1] + bin_edges[1:]) / 2

    table = pd.DataFrame({
        "i": np.arange(1, k + 1),
        "$[x_i, x_{i+1})$": [f"[{bin_edges[i]:.2f}, {bin_edges[i+1]:.2f})" for i in range(k)],
        "$\\tilde{x}_i$": np.round(mids, 2),
        "$m_i$": counts.astype(int)
    })

    h = (x_max - x_min) / k
    return table, h


n = 115
nu_tbl, h_nu = interval_table_hist(data_nu, n)
e_tbl, h_e   = interval_table_hist(data_e, n)

print("Исходные данные для nu:")
display(Markdown(nu_tbl[["i", "$\\tilde{x}_i$", "$m_i$"]].to_markdown(index=False)))
print(f"Шаг интервала h_nu = {h_nu:.4f}")

print("\nИсходные данные для E:")
display(Markdown(e_tbl[["i", "$\\tilde{x}_i$", "$m_i$"]].to_markdown(index=False)))
print(f"Шаг интервала h_e = {h_e:.4f}")

Исходные данные для nu:


|   i |   $\tilde{x}_i$ |   $m_i$ |
|----:|----------------:|--------:|
|   1 |          337.06 |       9 |
|   2 |          371.19 |       4 |
|   3 |          405.31 |      27 |
|   4 |          439.44 |      24 |
|   5 |          473.56 |      24 |
|   6 |          507.69 |      16 |
|   7 |          541.81 |       7 |
|   8 |          575.94 |       4 |

Шаг интервала h_nu = 34.1250

Исходные данные для E:


|   i |   $\tilde{x}_i$ |   $m_i$ |
|----:|----------------:|--------:|
|   1 |           72.18 |       5 |
|   2 |           87.54 |       7 |
|   3 |          102.91 |      16 |
|   4 |          118.27 |      25 |
|   5 |          133.63 |      30 |
|   6 |          148.99 |      22 |
|   7 |          164.36 |       7 |
|   8 |          179.72 |       3 |

Шаг интервала h_e = 15.3625


In [215]:
def conditional_variants(tbl: pd.DataFrame, h: float, name: str, *, digits_u: int = 3) -> dict:
    if h <= 0:
        raise ValueError(f"[{name}] h должен быть > 0, сейчас h={h}")

    df = tbl[["i", "$\\tilde{x}_i$", "$m_i$"]].copy()

    # условный нуль C: середина интервала с max частотой
    idx = int(df["$m_i$"].to_numpy().argmax())
    C = float(df.iloc[idx]["$\\tilde{x}_i$"])

    u = ((df["$\\tilde{x}_i$"].astype(float) - C) / h).round(digits_u)
    df["$u_i$"] = u

    # n_i * u_i^p (p=1..4)
    for p in (1, 2, 3, 4):
        col = "$n_i u_i$" if p == 1 else f"$n_i u_i^{p}$"
        df[col] = (df["$m_i$"] * (u ** p)).round(3)

    # n_i * (u_i + 1)^4
    df["$n_i (u_i+1)^4$"] = (df["$m_i$"] * ((u + 1) ** 4)).round(3)

    cols = [
        "i", "$\\tilde{x}_i$", "$m_i$", "$u_i$",
        "$n_i u_i$", "$n_i u_i^2$", "$n_i u_i^3$",
        "$n_i u_i^4$", "$n_i (u_i+1)^4$"
    ]
    out = df[cols].copy()

    # Σ строка
    sum_cols = ["$m_i$", "$n_i u_i$", "$n_i u_i^2$", "$n_i u_i^3$", "$n_i u_i^4$", "$n_i (u_i+1)^4$"]
    sigma = {
        "i": "Σ",
        "$\\tilde{x}_i$": "-",
        "$u_i$": "-",
    }
    for c in sum_cols:
        sigma[c] = float(out[c].sum())
    sigma["$m_i$"] = int(sigma["$m_i$"])

    out = pd.concat([out, pd.DataFrame([sigma])], ignore_index=True)

    n_total = int(df["$m_i$"].sum())

    print(f"\n=== {name}: условные варианты ===")
    print(f"h = {h:.4f} | C = {C:.2f} | n = {n_total}")

    return {"name": name, "h": h, "C": C, "n": n_total, "table": out}


def run_conditional(nu_tbl, h_nu, e_tbl, h_e) -> dict:
    res = {
        "nu": conditional_variants(nu_tbl, h_nu, "nu"),
        "E":  conditional_variants(e_tbl,  h_e,  "E"),
    }
    return res


cond = run_conditional(nu_tbl, h_nu, e_tbl, h_e)

nu_conditional_table = cond["nu"]["table"]
nu_C = cond["nu"]["C"]
nu_n = cond["nu"]["n"]

e_conditional_table = cond["E"]["table"]
e_C = cond["E"]["C"]
e_n = cond["E"]["n"]

print("\nТаблица условных вариант для nu:")
md(nu_conditional_table)

print("\nТаблица условных вариант для E:")
md(e_conditional_table)


=== nu: условные варианты ===
h = 34.1250 | C = 405.31 | n = 115

=== E: условные варианты ===
h = 15.3625 | C = 133.63 | n = 115

Таблица условных вариант для nu:


| i   | $\tilde{x}_i$   |   $m_i$ | $u_i$   |   $n_i u_i$ |   $n_i u_i^2$ |   $n_i u_i^3$ |   $n_i u_i^4$ |   $n_i (u_i+1)^4$ |
|:----|:----------------|--------:|:--------|------------:|--------------:|--------------:|--------------:|------------------:|
| 1   | 337.06          |       9 | -2.0    |         -18 |            36 |           -72 |           144 |                 9 |
| 2   | 371.19          |       4 | -1.0    |          -4 |             4 |            -4 |             4 |                 0 |
| 3   | 405.31          |      27 | 0.0     |           0 |             0 |             0 |             0 |                27 |
| 4   | 439.44          |      24 | 1.0     |          24 |            24 |            24 |            24 |               384 |
| 5   | 473.56          |      24 | 2.0     |          48 |            96 |           192 |           384 |              1944 |
| 6   | 507.69          |      16 | 3.0     |          48 |           144 |           432 |          1296 |              4096 |
| 7   | 541.81          |       7 | 4.0     |          28 |           112 |           448 |          1792 |              4375 |
| 8   | 575.94          |       4 | 5.0     |          20 |           100 |           500 |          2500 |              5184 |
| Σ   | -               |     115 | -       |         146 |           516 |          1520 |          6144 |             16019 |


Таблица условных вариант для E:


| i   | $\tilde{x}_i$   |   $m_i$ | $u_i$   |   $n_i u_i$ |   $n_i u_i^2$ |   $n_i u_i^3$ |   $n_i u_i^4$ |   $n_i (u_i+1)^4$ |
|:----|:----------------|--------:|:--------|------------:|--------------:|--------------:|--------------:|------------------:|
| 1   | 72.18           |       5 | -4.0    |         -20 |            80 |          -320 |          1280 |               405 |
| 2   | 87.54           |       7 | -3.0    |         -21 |            63 |          -189 |           567 |               112 |
| 3   | 102.91          |      16 | -2.0    |         -32 |            64 |          -128 |           256 |                16 |
| 4   | 118.27          |      25 | -1.0    |         -25 |            25 |           -25 |            25 |                 0 |
| 5   | 133.63          |      30 | 0.0     |           0 |             0 |             0 |             0 |                30 |
| 6   | 148.99          |      22 | 1.0     |          22 |            22 |            22 |            22 |               352 |
| 7   | 164.36          |       7 | 2.0     |          14 |            28 |            56 |           112 |               567 |
| 8   | 179.72          |       3 | 3.0     |           9 |            27 |            81 |           243 |               768 |
| Σ   | -               |     115 | -       |         -53 |           309 |          -503 |          2505 |              2250 |

In [216]:
def check_calc(cond_table: pd.DataFrame, n: int, name: str = "", *, tol: float = 1e-6) -> float:
    df = _no_sigma(cond_table)

    # суммы
    s1 = float(df["$n_i u_i$"].sum())
    s2 = float(df["$n_i u_i^2$"].sum())
    s3 = float(df["$n_i u_i^3$"].sum())
    s4 = float(df["$n_i u_i^4$"].sum())
    s_left = float(df["$n_i (u_i+1)^4$"].sum())

    s_right = s4 + 4*s3 + 6*s2 + 4*s1 + float(n)
    diff = abs(s_right - s_left)

    ok = "OK" if diff <= tol else "FAIL"

    print(f"=== Контроль {name} ===")
    print(f"LHS: Σ n_i (u_i+1)^4 = {s_left:.6f}")
    print(f"RHS: s4 + 4*s3 + 6*s2 + 4*s1 + n = {s_right:.6f}")
    print(f"Δ = {diff:.8f} -> {ok}")

    return diff


check_calc(nu_conditional_table, nu_n, "nu")
print()
check_calc(e_conditional_table, e_n, "E")

=== Контроль nu ===
LHS: Σ n_i (u_i+1)^4 = 16019.000000
RHS: s4 + 4*s3 + 6*s2 + 4*s1 + n = 16019.000000
Δ = 0.00000000 -> OK

=== Контроль E ===
LHS: Σ n_i (u_i+1)^4 = 2250.000000
RHS: s4 + 4*s3 + 6*s2 + 4*s1 + n = 2250.000000
Δ = 0.00000000 -> OK


0.0

Условные варианты для обоих признаков вычеслены, таблица заполнена, контроль вычислений прошел, что говорит о правильности вычислений.

In [217]:
def moments_from_cond_table(cond_table: pd.DataFrame, *, n: int, h: float, C: float, name: str = "") -> dict:
    df = _no_sigma(cond_table)

    # Σ n_i u_i^k
    s1 = float(df["$n_i u_i$"].sum())
    s2 = float(df["$n_i u_i^2$"].sum())
    s3 = float(df["$n_i u_i^3$"].sum())
    s4 = float(df["$n_i u_i^4$"].sum())

    inv_n = 1.0 / float(n)

    # M*_k
    M1, M2, M3, M4 = s1*inv_n, s2*inv_n, s3*inv_n, s4*inv_n

    # m_k (центральные)
    h2, h3, h4 = h*h, h*h*h, h*h*h*h
    m2 = h2 * (M2 - M1*M1)
    m3 = h3 * (M3 - 3*M1*M2 + 2*(M1**3))
    m4 = h4 * (M4 - 4*M1*M3 + 6*(M1*M1)*M2 - 3*(M1**4))

    return {
        "name": name, "n": int(n), "h": float(h), "C": float(C),
        "M1*": M1, "M2*": M2, "M3*": M3, "M4*": M4,
        "m1": 0.0, "m2": m2, "m3": m3, "m4": m4,
    }


def print_moments(m: dict) -> None:
    name = m.get("name", "")
    print(f"\n=== МОМЕНТЫ: {name} ===")
    print(f"M1* = {m['M1*']:.4f}, M2* = {m['M2*']:.4f}, M3* = {m['M3*']:.4f}, M4* = {m['M4*']:.4f}")
    print(f"m1 = {m['m1']:.0f}, m2 = {m['m2']:.4f}, m3 = {m['m3']:.2f}, m4 = {m['m4']:.0f}")


nu_moments = moments_from_cond_table(nu_conditional_table, n=nu_n, h=h_nu, C=nu_C, name="nu")
e_moments  = moments_from_cond_table(e_conditional_table,  n=e_n,  h=h_e,  C=e_C,  name="E")

print_moments(nu_moments)
print_moments(e_moments)


=== МОМЕНТЫ: nu ===
M1* = 1.2696, M2* = 4.4870, M3* = 13.2174, M4* = 53.4261
m1 = 0, m2 = 3348.1695, m3 = 8762.45, m4 = 29703148

=== МОМЕНТЫ: E ===
M1* = -0.4609, M2* = 2.6870, M3* = -4.3739, M4* = 21.7826
m1 = 0, m2 = 584.0110, m3 = -3098.78, m4 = 947347


In [218]:
def sample_stats(moments: dict, raw: np.ndarray, name: str, *, ddof: int = 0) -> dict:
    """
    x̄, D, σ:
      - по условным вариантам (через C, h, M1*, m2)
      - стандартно по raw
    ddof=0 -> смещённая, ddof=1 -> исправленная (если вдруг надо)
    """
    C = float(_pick(moments, "C", name=name))
    h = float(_pick(moments, "h", name=name))
    M1 = float(_pick(moments, "M1*", "M1_star", name=name))
    D_cond = float(_pick(moments, "m2", name=name))

    # через условные варианты
    xbar_cond = C + h * M1
    sigma_cond = float(np.sqrt(D_cond))

    # стандартно
    raw = np.asarray(raw, dtype=float)
    xbar_std = float(raw.mean())
    D_std = float(raw.var(ddof=ddof))
    sigma_std = float(raw.std(ddof=ddof))

    dx = abs(xbar_cond - xbar_std)
    dD = abs(D_cond - D_std)

    tag = "смещённые" if ddof == 0 else "исправленные"

    print(f"\n=== {name}: характеристики ({tag}) ===")
    print(f"cond: x̄ = {xbar_cond:.4f} | D = {D_cond:.4f} | σ = {sigma_cond:.4f}")
    print(f"std : x̄ = {xbar_std:.4f} | D = {D_std:.4f} | σ = {sigma_std:.4f}")
    print(f"Δx̄ = {dx:.6f} | ΔD = {dD:.6f}")

    return {
        "name": name,
        "x_bar_cond": xbar_cond, "D_cond": D_cond, "sigma_cond": sigma_cond,
        "x_bar_std": xbar_std,   "D_std": D_std,   "sigma_std": sigma_std,
        "dx_bar": dx, "dD": dD,
        "ddof": ddof,
    }


nu_sample_stats = sample_stats(nu_moments, data_nu, "nu", ddof=0)
e_sample_stats  = sample_stats(e_moments,  data_e,  "E",  ddof=0)


=== nu: характеристики (смещённые) ===
cond: x̄ = 448.6339 | D = 3348.1695 | σ = 57.8634
std : x̄ = 448.7217 | D = 3416.1834 | σ = 58.4481
Δx̄ = 0.087826 | ΔD = 68.013904

=== E: характеристики (смещённые) ===
cond: x̄ = 126.5499 | D = 584.0110 | σ = 24.1663
std : x̄ = 127.1296 | D = 583.8678 | σ = 24.1634
Δx̄ = 0.579674 | ΔD = 0.143192


In [219]:
def corrected_stats(sample_stats: dict, n: int, name: str) -> dict:
    D_b = float(_pick(sample_stats, "D_cond", "D", name=name))
    sigma_b = float(_pick(sample_stats, "sigma_cond", "sigma", name=name))

    k = n / (n - 1)
    s2 = k * D_b
    s = float(np.sqrt(s2))

    print(f"\n=== {name}: исправление дисперсии ===")
    print(f"n = {n} | k = n/(n-1) = {k:.6f}")
    print(f"D_b = {D_b:.4f} -> s² = {s2:.4f} (Δ={s2 - D_b:.4f})")
    print(f"σ_b = {sigma_b:.4f} -> s  = {s:.4f}")

    return {"name": name, "n": int(n), "k": k, "D_b": D_b, "sigma_b": sigma_b, "s2": s2, "s": s}


nu_corrected = corrected_stats(nu_sample_stats, nu_n, "nu")
e_corrected  = corrected_stats(e_sample_stats,  e_n,  "E")


=== nu: исправление дисперсии ===
n = 115 | k = n/(n-1) = 1.008772
D_b = 3348.1695 -> s² = 3377.5394 (Δ=29.3699)
σ_b = 57.8634 -> s  = 58.1166

=== E: исправление дисперсии ===
n = 115 | k = n/(n-1) = 1.008772
D_b = 584.0110 -> s² = 589.1339 (Δ=5.1229)
σ_b = 24.1663 -> s  = 24.2721


Исправленные оценки дисперсии и СКО ((s^2)) и ((s)) действительно немного больше смещённых $D_в$ и $\sigma_в$. Это полностью соответствует теории: замена деления на $n$ на деление на $n-1$ устраняет смещение и даёт несмещённую оценку генеральной дисперсии. В обоих случаях различие небольшое, потому что объём выборки $n=115$ достаточно велик, а множитель $\frac{n}{n-1}=1.008772$ близок к 1.

* Для признака **nu**: $D_в=3348.1695 \Rightarrow s^2=3377.5394$ $(прирост (\Delta=29.3699))$; $\sigma_в=57.8634 \Rightarrow s=58.1166$.
* Для признака **E**: $D_в=584.0110 \Rightarrow s^2=589.1339$ $(прирост (\Delta=5.1229))$; $\sigma_в=24.1663 \Rightarrow s=24.2721$.

Следовательно, для дальнейших статистических выводов корректнее использовать именно исправленные оценки $s^2$ и $s$, так как они являются несмещёнными оценками дисперсии и стандартного отклонения генеральной совокупности.


In [220]:
def skew_kurt(m: dict, name: str) -> dict:
    m2 = float(m["m2"])
    m3 = float(m["m3"])
    m4 = float(m["m4"])

    if m2 <= 0:
        raise ValueError(f"[{name}] m2 должен быть > 0, сейчас m2={m2}")

    sigma = float(np.sqrt(m2))
    As = m3 / (sigma ** 3)
    Ek = (m4 / (m2 ** 2)) - 3.0

    print(f"\n=== {name}: асимметрия/эксцесс ===")
    print(f"σ_в = {sigma:.4f}")
    print(f"As  = {As:.6f}")
    print(f"Ek  = {Ek:.6f}")

    return {"name": name, "sigma_b": sigma, "As": As, "Ek": Ek}


nu_shape = skew_kurt(nu_moments, "nu")
e_shape  = skew_kurt(e_moments,  "E")

nu_As, nu_Ek = nu_shape["As"], nu_shape["Ek"]
e_As, e_Ek   = e_shape["As"],  e_shape["Ek"]


=== nu: асимметрия/эксцесс ===
σ_в = 57.8634
As  = 0.045229
Ek  = -0.350356

=== E: асимметрия/эксцесс ===
σ_в = 24.1663
As  = -0.219563
Ek  = -0.222417


**Асимметрия (As):**

* Для признака **nu** коэффициент асимметрии **положительный и близок к нулю** $As \approx 0.045$. Это означает **очень слабую правостороннюю асимметрию**: правый “хвост” распределения чуть длиннее (есть немного значений больше среднего), но отклонение настолько маленькое, что распределение можно считать **почти симметричным**.

* Для признака **E** коэффициент асимметрии **отрицательный** $As \approx -0.220$. Это указывает на **умеренную левостороннюю асимметрию**: левый “хвост” длиннее, то есть встречаются значения модуля упругости **заметно ниже среднего**. При этом $|As| < 0.5$, поэтому асимметрия **не является сильной**.

**Эксцесс (Ek):**

* Для обоих признаков эксцесс **отрицательный** $Ek_{nu} \approx -0.350, Ek_{E} \approx -0.222$, следовательно распределения **плосковершинные** по сравнению с нормальным: вершина более “пологая”, а хвосты в целом **тоньше**.

* У признака **nu** плосковершинность выражена **сильнее**, чем у **E**, так как его эксцесс более отрицательный ((-0.350 < -0.222)).


In [221]:
def mode_median_variation(interval_tbl: pd.DataFrame, sample_stats: dict, n: int, name: str) -> dict:
    df = interval_tbl[["$[x_i, x_{i+1})$", "$m_i$"]].copy()
    df["L"], df["R"] = zip(*df["$[x_i, x_{i+1})$"].map(_parse_interval))
    df["freq"] = df["$m_i$"].astype(float)

    i_m = int(df["freq"].to_numpy().argmax())
    Lm = float(df.loc[i_m, "L"])
    h  = float(df.loc[i_m, "R"] - df.loc[i_m, "L"])
    fm = float(df.loc[i_m, "freq"])
    f0 = float(df.loc[i_m - 1, "freq"]) if i_m > 0 else 0.0
    f2 = float(df.loc[i_m + 1, "freq"]) if i_m < len(df) - 1 else 0.0

    denom = (fm - f0) + (fm - f2)
    Mo = Lm if denom == 0 else (Lm + h * ((fm - f0) / denom))

    df["cum"] = df["freq"].cumsum()
    half = n / 2.0
    i_med = int(df.index[df["cum"] >= half][0])
    Lmed = float(df.loc[i_med, "L"])
    Fprev = float(df.loc[i_med - 1, "cum"]) if i_med > 0 else 0.0
    fmed = float(df.loc[i_med, "freq"])
    hmed = float(df.loc[i_med, "R"] - df.loc[i_med, "L"])

    Me = Lmed + hmed * ((half - Fprev) / fmed)

    sigma = float(_pick(sample_stats, "sigma_cond", "sigma", name=name))
    xbar  = float(_pick(sample_stats, "x_bar_cond", "x_bar", name=name))
    V = (sigma / xbar) * 100.0 if xbar != 0 else np.nan

    print(f"\n=== {name}: Mo/Me/V ===")
    print(f"Mo* = {Mo:.4f}")
    print(f"Me* = {Me:.4f}")
    print(f"V*  = {V:.2f}%")

    return {"name": name, "Mo": Mo, "Me": Me, "V": V}


nu_mode_median_var = mode_median_variation(nu_tbl, nu_sample_stats, nu_n, "nu")
e_mode_median_var  = mode_median_variation(e_tbl,  e_sample_stats,  e_n,  "E")


=== nu: Mo/Me/V ===
Mo* = 418.4419
Me* = 447.2592
V*  = 12.90%

=== E: Mo/Me/V ===
Mo* = 131.8577
Me* = 128.2540
V*  = 19.10%


**Признак nu (объёмный вес):**

**Мода и медиана.** Мода $Mo^*\approx 418.44$ заметно меньше медианы $Me^*\approx 447.26$. Соотношение $Mo^* < Me^*$ обычно связывают с **правосторонней асимметрией** (положительной): наиболее частые значения находятся левее центра распределения, а “хвост” тянется вправо. Это в целом **согласуется** с ранее рассчитанным коэффициентом асимметрии для nu $As \approx 0.045$, который показывает **очень слабую** правостороннюю асимметрию (почти симметрия, но с лёгким смещением вправо).

**Коэффициент вариации.** $V^*\approx 12.90%$. Это значение лежит в интервале **10–20%**, значит вариация **средняя**: совокупность по признаку nu можно считать **достаточно однородной**, разброс относительно среднего умеренный.


**Признак E (модуль упругости):**

**Мода и медиана.** Мода $Mo^*\approx 131.86$ больше медианы $Me^*\approx 128.25$. Соотношение $Mo^* > Me^*$ характерно для **левосторонней (отрицательной) асимметрии**. Это **хорошо согласуется** с рассчитанным ранее $As \approx -0.220$: распределение имеет умеренно выраженный левый “хвост”, то есть присутствуют значения заметно ниже среднего.

**Коэффициент вариации.** $V^*\approx 19.10%$. Это тоже **средняя вариация**, но показатель близок к верхней границе 20%, значит неоднородность по E **более заметная**, чем по nu. Иными словами, признак **E варьирует сильнее**, чем nu.


## Выводы

В ходе выполнения работы были получены точечные оценки параметров распределения для двух количественных признаков: **nu** и **E** (на основе интервальных рядов и метода условных вариант).

Значения **выборочного среднего** $\bar{x}*{в}$, **выборочной дисперсии** $D*{в}$ и **среднеквадратического отклонения** $\sigma_{в}$, рассчитанные через **метод условных вариант**, согласованы с вычислениями по стандартным формулам, что подтверждает корректность построения интервальных таблиц и последующих расчётов.

**Исправленные оценки дисперсии и СКО** оказались немного больше смещённых, что соответствует теории для конечных выборок, поскольку $\frac{n}{n-1}>1$. При $n=115$ множитель $k=\frac{n}{n-1}=1.008772$ близок к 1, поэтому разница небольшая:

* для **nu**: $D_{в}=3348.1695 \Rightarrow s^2=3377.5394$ $\Delta=29.3699$, $\sigma_{в}=57.8634 \Rightarrow s=58.1166$;
* для **E**: $D_{в}=584.0110 \Rightarrow s^2=589.1339$ $\Delta=5.1229$, $\sigma_{в}=24.1663 \Rightarrow s=24.2721$.
  Для дальнейших статистических выводов целесообразно использовать именно $s^2$ и $s$ как несмещённые оценки.

По форме распределений:

* для **nu** коэффициент асимметрии $As=0.045229$ положителен, значит наблюдается **очень слабая правосторонняя асимметрия**; эксцесс $Ek=-0.350356$ отрицателен, следовательно распределение **плосковершинное** (менее “острое”, чем нормальное).
* для **E** $As=-0.219563$ отрицателен, что указывает на **умеренную левостороннюю асимметрию**; $Ek=-0.222417$ также отрицателен — распределение **плосковершинное**.

**Коэффициенты вариации** составили:

* для **nu**: $V^*=12.90%$,
* для **E**: $V^*=19.10%$.
  Оба значения заметно меньше 33%, значит совокупность можно считать **однородной** по обоим признакам, а выборочные средние — **достаточно надёжными** характеристиками центра распределения.

Дополнительно рассчитанные **мода** и **медиана** по интервальным рядам уточняют центральную тенденцию:

* для **nu**: $M_o^*=418.4419$, $M_e^*=447.2592$;
* для **E**: $M_o^*=131.8577$, $M_e^*=128.2540$.

  Эти показатели дополняют среднее значение и помогают описать распределение более полно, особенно при наличии асимметрии.
